### ISSUE TITLE
Attempt to work around GCC 6's overzealous optimizations

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/UXP#14]
`nsIPresShell::AllocateFrame` and `::AllocateByObjectID` implement a pool allocator that zeroes out returned memory. This allocator is in place so that we don't have to worry about uninitialized variables in related objects.

Unfortunately this falls foul of GGC 6's `lifetime-dead-store-elimination` optimisations.  This detects
the `memset(..)` as occurring _before_ the lifetime of the allocated object begins, and so removes the call in its infinite wisdom. This in turn causes `nsFrame` and many child classes to have uninitialized fields after construction (normally zeroed out), which leads to multiple segfaults in short order.

Removing the use of this zeroing allocator is an option, but risky, because then we'd have to go through all the consumers of it and manually initialize the variables.

**Comment by trav90** [comment]
As a temporary workaround, we could disable lifetime-dse optimizations when using GCC 6 as we currently do for Pale Moon:

http://xref.palemoon.org/palemoon-trunk/source/configure.in#1649

**Comment by wolfbeast** [comment]
Sounds like a plan.

Seems like Mozilla removed the zeroing allocator in [Bug 1316556](https://bugzilla.mozilla.org/show_bug.cgi?id=1316556) - Remove zeroing allocation in class nsIPresShell which has a bunch of expected dependencies (crashes and sec issues caused by doing that because stuff no longer gets initialized on _any_ compiler, not just the GCC6 "optimized" one)
If we're going to follow that, which I'm still on the fence about because it's a compiler bug we can work around, then we must do all of this work and scour bugzilla for any other non-linked bugs to the same effect.

**Comment by trav90** [comment]
Looks like we are already disabling lifetime-dse optimizations with any version of GCC 5 or above:

http://xref.palemoon.org/uxp-trunk/source/build/autoconf/compiler-opts.m4#182

**Comment by wolfbeast** [comment]
So, for a workaround we don't need to do anything, IIUC?

**Comment by trav90** [comment]
Exactly. Our tree already conatins the workaround which disables lifetime-dse optimizations when using GCC versions 5 or above (tested and confirmed that this is the case).

**Comment by mattatobin** [comment]
So is this `RESOLVED INVALID` or what?

**Comment by trav90** [comment]
Depends on if we are going to remove the zeroing allocator as Mozilla did or continue working around the problem by disabling lifetime-dse optimizations in modern GCC versions.

**Comment by wolfbeast** [comment]
The question is: is disabling this optimization going to hurt us, and if so, to what extent? I'd prefer to keep the allocator because it is a good safeguard against potential sec issues due to uninitialized use.

**Comment by trav90** [comment]
OK, so I've been doing a bit of research on this. As far as I can tell lifetime-dse optimization was introduced in GCC 4.9 and is used with -O1 and above. Beginning with GCC 6, this was changed to be much more aggressive (so-called "level 2 dse optimization") and is still enabled by default at -O1 and above. However, due to the backlash from so many projects experiencing segfaults and other issues caused by the more aggressive optimization, a `-flifetime-dse=1` option was added which restores the previous behavior from GCC 4.9 and 5.

Mozilla basically did not want to deal with this at all (as they were working on removing the zeroing allocator for FF53), so they just completely disabled all lifetime dse optimizations for esr-52 and moved on.

So as I see it we have three options:
          1. Leave it as-is (lifetime-dse optimization disabled completely for GCC 5 and above).
          2. Force GCC 6 and above to use "level 1" optimization used in earlier versions.
          3. Fix/remove the zeroing allocator and allow GCC to use default options.

I plan to build using the level 1 optimization and compare it to a build with it disabled to see if there is any discernible (dis)advantages to using it or not.

**Comment by wolfbeast** [comment]
Option 2 is the preferred option if it is stable.
I would greatly prefer not using option 3 because it is a very smart solution that provides a lot of defense in depth on all platforms we build on, and I'd rather not lose that because of a compiler issue on just one.

So in order of preference: 2 if possible, otherwise 1, and only if it can't be helped otherwise go for 3.

**Comment by khronosschoty** [comment]
Built Basilisk from the branch "buildsystem-work" in Travis repo.
Verefied in about:buildconfig that the "-flifetime-dse=1" flag was
present. Testing out Basilisk now, so far so good.

**Comment by mattatobin** [comment]
Push up a branch and link it I will take a gander @Justoff

**Comment by trav90** [comment]
Going to continue testing for a bit but so far so good here with Pale Moon and -flifetime-dse=1.

**Comment by mattatobin** [comment]
I would argue Basilisk is more important test wise cause it has working drm and webrtc built.

**Comment by trav90** [comment]
Agreed and I also will be testing Basilisk but I also wanted to give Pale Moon a quick test as well.

**Comment by khronosschoty** [comment]
Still testing Basilisk out and working great, no issues... so it makes
sense to test out Pale Moon as well. I'll take Pale Moon tomorrow.

**Comment by khronosschoty** [comment]
Seems ever so slightly faster, but I can't be totally sure. It is for
sure no slower.

**Comment by trav90** [comment]
My testing confirms @khronosschoty's results. Submitting PR.
